src/main/java/macrobase/analysis/classify/EWAppxPercentileOutlierClassifier.java

```diff
@@ -0,0 +1,104 @@
+             conf.getInt(MacroBaseConf.WARMUP_COUNT, MacroBaseDefaults.WARMUP_COUNT),
+             conf.getDecayType(),
+             conf.getDouble(MacroBaseConf.SUMMARY_UPDATE_PERIOD, MacroBaseDefaults.SUMMARY_UPDATE_PERIOD),
+             conf.getDecayType(),
+             conf.getDouble(MacroBaseConf.SUMMARY_UPDATE_PERIOD, MacroBaseDefaults.SUMMARY_UPDATE_PERIOD),
+             conf.getDouble(MacroBaseConf.DECAY_RATE, MacroBaseDefaults.DECAY_RATE),
+             conf.getDouble(MacroBaseConf.TARGET_PERCENTILE, MacroBaseDefaults.TARGET_PERCENTILE));
+    }
+
+    private void updateThreshold(double percentile) {
+        List<DatumWithNorm> norms = reservoir.getReservoir();
+        norms.sort((a, b) -> a.getNorm().compareTo(b.getNorm()));
+        currentThreshold = norms.get((int)(percentile * norms.size())).getNorm();
+    }
+
+    public EWAppxPercentileOutlierClassifier(MacroBaseConf conf,
+                                             Iterator<Datum> input,
+                                             int sampleSize,
+                                             int warmupCount,
+                                             MacroBaseConf.PeriodType updatePeriodType,
+                                             double updatePeriod,
+                                             MacroBaseConf.PeriodType decayPeriodType,
+                                             double decayPeriod,
+                                             double decayRate,
+                                             double percentile) {
+        super(conf, input);
+        reservoir = new ExponentiallyBiasedAChao<>(sampleSize, decayRate, conf.getRandom());
+
+        this.percentileUpdater = new Periodic(updatePeriodType,
+                                              updatePeriod,
+                                              () -> updateThreshold(percentile));
+
+        this.reservoirDecayer = new Periodic(decayPeriodType,
+                                             decayPeriod,
+                                             reservoir::advancePeriod);
+
+        List<DatumWithNorm> warmupInput = new ArrayList<>(warmupCount);
+        for(int i = 0; i < warmupCount; ++i) {
+            Datum d = input.next();
+            DatumWithNorm dwn = new DatumWithNorm(d, d.getMetrics().getNorm());
+            warmupInput.add(dwn);
+            reservoir.insert(dwn);
+            reservoirDecayer.update();
+            percentileUpdater.update();
+        }
+        updateThreshold(percentile);
+
+        for(DatumWithNorm dwn : warmupInput) {
+            warmupOutput.add(new OutlierClassificationResult(dwn.getDatum(), dwn.getNorm() > currentThreshold));
+        }
+    }
+
+    @Override
+    public boolean hasNext() {
+        return input.hasNext();
+    }
+
+    @Override
+    public OutlierClassificationResult next() {
+        if(!warmupOutput.isEmpty()) {
+            return warmupOutput.remove(0);
+        }
+
+        reservoirDecayer.update();
```

Comment:
I presume these aren't actually decaying / updating percentiles every tuple. If this is indeed the case, consider renaming to `updateIfNecessary()`.